File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        
        ListNode runner = dummy;
        ListNode walker = dummy;

        for(int i = 0; i <= n; i++)  {
            runner = runner.next;
        }

        while (runner != null)  {
            walker = walker.next;
            runner = runner.next;
        }
        
        walker.next = walker.next.next;
        return dummy.next;
    }
}

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # Initialize two pointers:
        # 'slow' moves one step at a time
        # 'fast' moves two steps at a time
        slow = head
        fast = head

        # Traverse the list while fast pointer can move
        # If fast reaches the end → no cycle
        while fast and fast.next:

            # Move pointers
            fast = fast.next.next   # moves 2 steps
            slow = slow.next        # moves 1 step

            # If there is a cycle, fast will eventually meet slow
            if fast == slow:
                return True

        # If we exit the loop, fast reached the end → no cycle
        return False